### PR TITLE
Add a python file for predictor data prepare

### PR DESCRIPTION
**Pull request by umdsgy**

### CODE REVIEW

**Review by pvojnov — COMMENTED** [review]
This pull request is very similar to another one (https://github.com/BigDataWUR/AgML-CY-Bench/pull/213), not sure why there are 2 since they use pretty much the same code.
I commented most of the things in (https://github.com/BigDataWUR/AgML-CY-Bench/pull/213), here I commented only the differences.

File: data_preparation/AgML_predictor_data_prepare.py [hunk 1]
```diff
@@ -0,0 +1,1074 @@
...
+        classification_out['values'] = [i['val_count'] for i in class_res]
+        output['classification'] = classification_out
+
+    return output
+
+
+"""
+@author: Abdelrahman
+""" 
+def netCDFToTiff(netcdf_file, in_dir, default_crs='EPSG:4326'):
```
[the remark this answers is not in the document]
Comment (umdsgy) [reply data_preparation/AgML_predictor_data_prepare.py:568]: It is not used in the predictor preparation. So just keep it right here with your agreement.

**Comment by krsnapaudel** [comment]
@umdsgy : Please address the comments from #213 as well. Thanks.

File: data_preparation/AgML_predictor_data_prepare.py [hunk 2]
```diff
@@ -0,0 +1,1077 @@
...
+def arr_stats(arr, weights=None,
+              output=('min', 'max', 'sum', 'mean', 'count', 'mode')):
+    """
+    Extracts statistics from input array (arr).
+    It uses weights array as weights if provided.
+    List of statistics to extract is defined in the output parameter as a list or a comma separated string.
+    :param arr: Input array
+    :param weights: Array providing weights for each pixel value. Used to calculate stats.
+    :param output: List of values to extract, can be a list or comma separated string.
+        Possible values are:
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:239]: Missing description for output option "mode"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 3]
```diff
@@ -0,0 +1,1077 @@
...
+        else:
+            ind = np.isnan(_arr)
+            out_vals['mean'] = np.ma.mean(_arr[~ind])
+
+    if 'std' in _output:
+        arr_size = np.size(arr_compressed)
+        if weights is not None:
+            # combine mask from the arr and create a compressed arr
+            weights_compressed = np.ma.array(weights, mask=_arr.mask).compressed()
+            ind = np.isnan(arr_compressed) | np.isnan(weights_compressed) | (_arr <= -9999)
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:276]: These arrays might not be of the same shape, you are mixing compressed and uncompressed arrays

File: data_preparation/AgML_predictor_data_prepare.py [hunk 4]
```diff
@@ -0,0 +1,1077 @@
...
+    
+    """
+    Reads the data fom the raster file and returns it as a numpy masked array.
+    The geometry is used to define the mask.
+    It returns only a subset of the data.
+        It reads only the raster window containing the geometry. This optimizes the performance by reducing disk reads.
+    The "use_pixels" parameter can be used to define how to handle the border pixels.
+        Some require shapely to be available on the system
+    :param ds: Raster file path or an instance of a rasterio.DatasetReader (already opened raster file).
+    :param mask: geojson geometry used to mask the array, mask is a read mask (only data covered by the geometry is used)
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:88]: The description should be updated to match the code change from the original. "mask" changed form geometry -> iterable over geometries

File: data_preparation/AgML_predictor_data_prepare.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pvojnov) [reply data_preparation/AgML_predictor_data_prepare.py:568]: Sure no problem, it was just an observation if it was left here unintentionally

File: data_preparation/AgML_predictor_data_prepare.py [hunk 5]
```diff
@@ -0,0 +1,1077 @@
...
+}
+
+# Angola (AO), Burkina Faso (BF), Ethiopia (ET), Lesotho (LS), Madagascar (MG), Malawi (MW),
+# Mozambique (MZ), Niger (NE), Senegal (SN), Chad (TD), South Africa (ZA), Zambia (ZM)
+FEWSNET_COUNTRIES = ["AO", "BF", "ET", "LS", "MG", "MW", "MZ", "NE", "SN", "TD", "ZA", "ZM"]
+FEWSNET_ADMIN_ID_KEY = "adm_id"
+
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:67]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 6]
```diff
@@ -0,0 +1,1077 @@
...
+"""
+
+class UnableToExtractStats(Exception):
+    pass
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:74]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 7]
```diff
@@ -0,0 +1,1077 @@
...
+                           transform=out_transform) as tmp_dataset:
+            tmp_dataset.write(input_geom_mask.astype(np.uint8), 1)
+
+    # mask data arrays
+    source = np.ma.array(source, mask=input_geom_mask)
+
+    return source
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:158]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 8]
```diff
@@ -0,0 +1,1077 @@
...
+    :return: rasterio.windows.Window
+    """
+    _col_off = round(window.col_off)
+    _row_off = round(window.row_off)
+    _width = round(window.col_off + window.width) - _col_off
+    _height = round(window.row_off + window.height) - _row_off
+    return rasterio.windows.Window(_col_off, _row_off, _width, _height)
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:175]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 9]
```diff
@@ -0,0 +1,1077 @@
...
+            min_res_window_size = rasterio.windows.shape(ds_window)
+
+    # new bounds fitted to the highest resolution file
+    out_bounds = max_res_ds.window_bounds(round_window(max_res_ds.window(*max_extent_window_bounds)))
+    out_shape = rasterio.windows.shape(round_window(max_res_ds.window(*out_bounds)))
+
+    return out_bounds, out_shape
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:228]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 10]
```diff
@@ -0,0 +1,1077 @@
...
+        out_vals['weight_sum'] = weights_compressed.sum()
+
+    # convert to regular py types from np types which can cause problems down the line like JSON serialisation
+    out_vals = {k: v.item() for k, v in out_vals.items()}
+
+    return out_vals
+
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:328]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 11]
```diff
@@ -0,0 +1,1077 @@
...
+            cls['val_count'] = np.sum(np.logical_and(arr >= cls['min'], arr <= cls['max']) * _weights)
+        else:
+            raise ValueError('Parameter "border_include" not defined properly. Allowed values are "min", "max", "both" or None')
+        cls_out.append(cls)
+
+    return cls_out
+
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:361]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 12]
```diff
@@ -0,0 +1,1077 @@
...
+    if scale and scale != 1:
+        arr = arr * scale
+    if offset and offset != 0:
+        arr = arr + offset
+
+    return arr
+
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:386]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

File: data_preparation/AgML_predictor_data_prepare.py [hunk 13]
```diff
@@ -0,0 +1,1077 @@
...
+    elif callable(nodata):
+        return np.ma.array(arr, mask=nodata(arr))
+    elif type(nodata) in [list, tuple]:
+        return np.ma.array(arr, mask=np.isin(arr, nodata))
+    else:
+        return np.ma.array(arr, mask=arr == nodata)
+
+
+"""
+@author: Joint Research Center
```
Comment (pvojnov) [inline data_preparation/AgML_predictor_data_prepare.py:407]: Please change the author to  "Joint Research Centre - D5 Food Security - ASAP"

**Review by pvojnov — COMMENTED** [review]
I added all the comments from the PR 213.
I was sure I commented the authorship last time but now I see I didn't. Now I've added authorship comment on all JRC occurrences.

Bellow is  general comment that I made on PR 213:
_I see you didn't use our code for multiprocessing. But instead you are multiprocessing over multiple files.
Depending on hardware, number of spatial units and raster indicators this approach might run significantly slower since it reads multiple files at once and the disk IO might be a bottleneck.
If multiprocessing over geometries on each file you read one file at a time and extract multiple geometries at once, mostly CPU intense. That is why we used this approach in our code._

File: data_preparation/AgML_predictor_data_prepare.py (hunk 4 above)
Comment (umdsgy) [reply data_preparation/AgML_predictor_data_prepare.py:88]: Hi Pvojnov and Dilli, I do not have time to test the new version (which iterable over geometries). Please use this version.

Comment (krsnapaudel) [reply data_preparation/AgML_predictor_data_prepare.py:88]: @umdsgy : The idea is to review, address and test. If you cannot test, then the code will stay as a pull request and never merged to main.

Comment (pvojnov) [reply data_preparation/AgML_predictor_data_prepare.py:88]: @umdsgy, this comment was just to update the function description to match the changes you made to the original code.
@krsnapaudel, I believe that  the code in this PR is probably tested and that @umdsgy is referring to my general comment to try to use the additional part of the code that was shared after there was some complaints that the execution is slow.

Comment (umdsgy) [reply data_preparation/AgML_predictor_data_prepare.py:88]: yes, @pvojnov is on the same page with me. The current code has been tested and run for months.

Comment (krsnapaudel) [reply data_preparation/AgML_predictor_data_prepare.py:88]: @umdsgy: If it's tested, please share some results from the run. They need to match the format of data in the benchmark dataset.

File: data_preparation/AgML_predictor_data_prepare.py (hunk 2 above)
Comment (pvojnov) [reply data_preparation/AgML_predictor_data_prepare.py:239]: @umdsgy, @krsnapaudel after another colleague asked for something similar I realised that this implementation doesn't take into account the AFI. The output would be correct only if it's done with no AFI provided.

**Review by pvojnov — COMMENTED** [review]
@umdsgy, @krsnapaudel I made an additional comment for the mode output not sure if you managed to see it.
The comment below:

> after another colleague asked for something similar I realised that this implementation doesn't take into account the AFI. The output would be correct only if it's done with no AFI provided.

**Comment by ritviksahajpal** [comment]
@pvojnov , can you clarify your comment. Why is the output incorrect if an AFI is provided? Do you have a version of the code which works when the AFI is provided?

**Comment by pvojnov** [comment]
> @pvojnov , can you clarify your comment. Why is the output incorrect if an AFI is provided? Do you have a version of the code which works when the AFI is provided?

Because I believe the current code finds the most frequent value but if you have AFI then you need to take it into account as weights.
Basically for each value you should look at it's sum of AFI values and then the one with the highest sum would be the desired result.

In _geom_extract_ function we have the _classification_ variable that finaly uses the _arr_classes_count_ function that can to some extent do a similar thing. It takes into account the AFI.
For it you define the boundary values and the output is either count of pixels within the boundaries or AFI sum if AFI is provided.

**Review by pvojnov — COMMENTED** [review]
@umdsgy this should be correct if using AFI but I believe you should get an error if no AFI is provided.
Also this approach might be a bit slow since you are looping through all elements of an array.

I had a look at the _np.bincount_ and I think you might use the original approach, slightly adapted.
_np.bincount_ has a weights parameter that looks like it does what we need.
Maybe something to test since it should be a lot faster then looping through individual values.

Also if using _np.bincount_ I think you need to handle cases if the indicator array is float, since this function only accepts ints.

**Comment by krsnapaudel** [comment]
@umdsgy: Please create a new branch, pull latest changes from main, add your file, address the last remaining comments from Petar and update the pull request.

**Comment by krsnapaudel** [comment]
Tracked by #264

**Pull request closed without merging**